### ISSUE TITLE
Code Generation - Loose input structure option leads to broken code

### ISSUE THREAD

**Issue by bmaclach** [JacquesCarette/Drasil#1453]
Choosing "Loose" as the input structure for code generation (as opposed to "AsClass") leads to very broken code. The differences between Loose and AsClass seem to be the following:

- Loose combines the declaration of input variables, checking input_constraints, and calculating derived values into a single module whereas AsClass has separate modules for each
- Loose does not have a class for storing all of the input variables. It simply declares/defines different variables for each input. 

A lot of our generator assumes that there is an "InputParameters" object and that the DerivedValues and InputConstraints modules exist, but these assumptions do not hold when "Loose" is chosen (see related issue #1177).

I don't know how best to fix this because I don't know what the intention behind the "Loose" option was. @JacquesCarette do you know what code generated with this option was supposed to look like? Is the idea that all the input variables would be global? Or local variables that get passed to all of the functions that use them?

**Comment by smiths** [comment]
@bmaclach, this issue reminds me of our recent discussions with @samm82 and @JacquesCarette.  We want implementation variabilities in our generated code.  We could have code that follows the human's design, or code that ignores the imposed design structure.  I think the problem with AsClass and Loose is that they aren't really defined.  The exercise with Projectile should help us get a better idea of the design variabilities we want to control.

**Comment by JacquesCarette** [comment]
I think the idea was that `AsClass` (really `AsRecord`) should pass a single record for all inputs, while `Loose` (really `Unbundled`) passes all variables explicitly, "unbundled" as it were.

The code should work equally well either way. That, of course, doesn't mean that the current generator is set up to deal with it in either configuration!

**Comment by bmaclach** [comment]
For the `Loose` option, passing each input explicitly to the functions that use them is now working. But one thing I still don't understand about the `Loose` option is how it's supposed to work with setting the values of the input variables. The `func_get_input` function (i.e. the function that reads an input file and sets the values) would have to return multiple values if the inputs are all unbundled, which we can't currently do with GOOL (unless we use a class, but the point of `Loose` seems to be to not use a class). Should we just inline the functionality of `func_get_input` in our Control module? @JacquesCarette @smiths

**Comment by smiths** [comment]
I agree with your proposal to inline the functionality into the control module.

If I were writing the code that we are generating, I would likely go with the AsRecord (AsClass) option.  A separate module for reading the data can be written with a simple interface and provides a nice separation of concerns.  However, if we go with the Unbundled (Loose) option, I agree that there wouldn't be much point in creating an Input Format Hiding Module.  It helps people, but it doesn't matter for the generated code.  The Control module can be "ugly" and long, but it doesn't matter for the generated code.

**Comment by JacquesCarette** [comment]
First thing: I think we should rename `Loose` to `Unbundled` and `AsClass` to `Bundled`.

For `func_get_input`, I think the usual implementation of passing many outputs in a language in an 'unbundled' way is to provide them all as writeable inputs. I think most of our languages allow that. If a language does **not**, then I would say that asking for `Unbundled` as a global option should be an error. Note that, for example, Python allows multiple outputs, so `Unbundled` in Python should use that feature.

inlining or not, to me, is a completely different, orthogonal issue.

I also disagree with @smiths regarding the usefulness of `Unbundle` and an Input Format Hiding Module. For example, the IFH module can hide the different between (say) inputs stored in csv, little-endian binary and big-endian binary formats, for example. And I can come up with other examples where the IFH Module will have a secret that it will keep.

**Comment by smiths** [comment]
@JacquesCarette, I agree with you that Input Format Hiding Modules are possible with `Unbundled`.  I was thinking more from the point of view that for the generated code we can relax the "one module one secret" rule.  The control module could then have two secrets: the input format and the overall control.  I don't like that design, but it is a fairly common design for scientific software.  I thought `Unbundled` would let us capture that pattern, but I agree with you now that there are actually two separate variations here.  `Unbundled` versus `Bundled' and `Inline` versus `NotInline`.  For inlining we will likely need to break it down into further potential variations.

**Comment by bmaclach** [comment]
@JacquesCarette After looking into this, Java doesn't have a built-in way to pass parameters by reference. It seems like the way to do it is to make a class for each parameter, then pass them as objects. (sources: https://stackoverflow.com/questions/5614562/how-to-do-the-equivalent-of-pass-by-reference-for-primitives-in-java, https://stackoverflow.com/questions/1068760/can-i-pass-parameters-by-reference-in-java). Is that a good enough solution that we'd want to implement it that way, or still throw the error?

Right now all of our options are "global" options, right? We don't have a way to choose one option for Python, and a different one for Java, for example. But if someone wasn't targeting Java and wanted `Unbundled` inputs, we should allow that, shouldn't we?

Also, I'm wondering how we would have Python return multiple outputs while other languages use writable inputs, presumably from the same GOOL code? Does this go hand-in-hand with what you said at a previous meeting about how parts of GOOL are too low-level right now, and it should offer more higher-level functions? So for this case, the high-level description of what we want is to modify many values in a single function, but how to accomplish that (either by returning many outputs or passing inputs by reference) is a low-level detail that is abstracted away in the GOOL interface?

**Comment by JacquesCarette** [comment]
I'm undecided about Java. Opinions?

Yes, currently all our options are global (and yes, we'll eventually want to allow local options). Each generator gets passed its own Options record, so there ought to be a way to allow that. It's harder in the current test suite which tries to be thorough. So it depends on the usage scenario you have in mind.

Yes, this is one example of GOOL being too low-level: if we had things marked as input, other as output, then we would have more freedom to adjust by language.

**Comment by bmaclach** [comment]
I'm leaning towards not allowing this in Java. I did a small experiment to see how this would work, and all of these extra classes we'd have to generate to imitate pass-by-reference parameters would each have to exist in their own file (I'm not a Java expert, so maybe there's a way to avoid this that I'm not aware of). So every time we call a function that needs pass-by-reference parameters, we are generating one more file per parameter. This seems bad. Here's what it would look like for a very simple function call with one pass-by-reference parameter:

In A.java:
```
public class A {
  public int a;
  public A(int a') throws Exception {
    a = a';
  }
}
```
In SomeModule.java:
```
public static void setToFive(A a) throws Exception {
    A.a = 5;
}
```
In Control.java:
```
int a = 4; // a could be set to some value already, before calling setToFive
A a_temp = new A(a); // Initialize class instance
SomeModule.setToFive(a_temp);
a = a_temp.a; // Extract value from class instance and assign it to original a, so that a can be used regularly in future code
```
Which just seems like a lot of mess for something so simple (I realize that a lot of mess is less of a big deal with generated code, though). Also, our examples would never be this simple because we would only be doing this for cases where we have multiple pass-by-reference parameters.

**Comment by JacquesCarette** [comment]
You don't have to have a setter and getter - you can just have
```java
public class A { public int a; }
```
as the declaration of a single-value 'record', which is then mutable.  In fact, with generics, you might be able to do 
```java
public class GenericRecord<T> {public T t;}
```
instead.  Which isn't too horrifying.

**Comment by bmaclach** [comment]
Also - the generic class would mean we only ever need to generate one extra file, which is great.

**Comment by JacquesCarette** [comment]
We do get less type safety if we pass things by reference more pervasively - though in some sense, because of all the languages and we're doing the generation, I'm not sure that's a huge danger.

But maybe it is best to disallow multiple outputs in Java in Unbundled mode.

**Comment by bmaclach** [comment]
It's just occurred to me that there may be a much cleaner solution for Java. Rather than trying to emulate pass-by-reference parameters, what if we go back to the original idea of returning multiple outputs? If we have multiple outputs, could we just put those outputs into a list of `Object`s and return that list, then assign each list member to the correct variable? (Every class in Java is an instance of `Object`, so we'd just have to cast each list member back to its original type when we extract from the list, which is easy).

**Comment by JacquesCarette** [comment]
It's massively untyped, but it would work. I'd kind of prefer a vector over a list, at least we could have its length be static. Good thinking though.

**Comment by bmaclach** [comment]
@JacquesCarette My understanding is that we will make the "type" of list (i.e. Vector, ArrayList, etc.) a choice for the user of Drasil. Would we want to just use whichever type of list they've chosen, or always use vector for this specific case?

**Comment by JacquesCarette** [comment]
To me, this is an internal choice for doing multiple output, which is different than an 'implementation choice' that a user might meaningfully make. Or at least, that's how I currently see it.

**Comment by bmaclach** [comment]
@JacquesCarette Sounds good, though I don't think that a Vector in Java has static length. Did you mean an array, i.e. something declared like `Object[] outs = new Object[10]`?

**Comment by JacquesCarette** [comment]
Yes, array. I tend to forget the exact name of data-structures in each language...

**Comment by bmaclach** [comment]
`Loose` (now `Unbundled`) option is fixed as of merges of #1627 and #1630, this issue can be closed.
